src/firestore/firestore.ts

```diff
@@ -258,6 +258,75 @@ export function subscribeToClientUpdatesV2(
   );
 }
 
+/**
+ * This will migrate legacy data to the new subcollection, if there is no document in the
+ * new subcollection but legacy data exists. It will not change any data in legacy collections,
+ * but the expectation is that they will be ignored once the new collection is populated.
+ * TODO(#2108): Remove this function after migration is complete
+ */
+async function migrateOpportunityUpdate(
+  updateDocRef: DocumentReference,
+  clientId: string,
+  clientRecordId: string,
+  opportunityType: OpportunityType
+) {
+  // if destination document does not already exist, we will look for a legacy document to migrate
+  let dataToMigrate: OpportunityEdits | undefined;
+  if (!(await getDoc(updateDocRef)).exists()) {
+    const {
+      docRef: v2UpdatesDocRef,
+      oldDocument: v1UpdatesDoc,
+    } = await getClientUpdatesV2DocRef(clientId, clientRecordId);
+
+    // legacy format(s): object nested directly in the update doc
+    let legacyRecord = v1UpdatesDoc;
+    // this doc will only be returned if the v2 doc does not exist
+    if (!v1UpdatesDoc) {
+      legacyRecord = (await getDoc(v2UpdatesDocRef)).data();
+    }
+
+    dataToMigrate = legacyRecord?.[opportunityType];
+  }
+
+  if (dataToMigrate) {
+    // write old + new data to new destination
+    setDoc(updateDocRef, { ...dataToMigrate });
```

Comment:
I'm not sure if you're making a copy on purpose, but wanted to flag this just in case.

```suggestion
    setDoc(updateDocRef, dataToMigrate);
```